[Merged by Bors] - feat(SetLike): add exists_not_mem_of_ne_top lemma

jessealama opened the pull request:
Adds a bridging lemma for `SetLike` structures: if `s ≠ ⊤` and the top element coerces to the universal set, then there exists an element not in `s`.

This generalizes a pattern that was specific to `AffineSubspace` and makes it available for all `SetLike` structures (submodules, subalgebras, etc.) where `⊤ = univ`.

Extracted from #30854 per reviewer feedback: https://github.com/leanprover-community/mathlib4/pull/30854#discussion_r2462326156

---

**Potential reviewers** (recent contributors to `Mathlib/Data/SetLike/Basic.lean`):
- @kim-em 
- @themathqueen

jessealama committed:
feat(SetLike): add exists_not_mem_of_ne_top lemma

j-loreaux commented on Mathlib/Data/SetLike/Basic.lean:
```diff
@@ -243,4 +243,13 @@ lemma mem_of_subsingleton {A F} [Subsingleton A] [SetLike F A] (S : F) [h : None
+lemma exists_not_mem_of_ne_top (s : S) (hs : s ≠ ⊤) (h_top : ((⊤ : S) : Set A) = Set.univ) :
+    ∃ a : A, a ∉ s := by
+  have : (s : Set A) ≠ Set.univ := fun h_eq => hs (coe_injective (h_eq.trans h_top.symm))
+  exact (Set.ne_univ_iff_exists_notMem _).mp this
```
I'm not sure how useful this is in practice, but to make it more so, you can let `h_top` have an autoparam.
```
lemma exists_not_mem_of_ne_top (s : S) (hs : s ≠ ⊤)
    (h_top : ((⊤ : S) : Set A) = Set.univ := by simp) :
    ∃ a : A, a ∉ s := by
  simpa [-SetLike.coe_set_eq, SetLike.ext'_iff, h_top, Set.ne_univ_iff_exists_notMem] using hs
```

jessealama replied:
Yeah, this is admittedly a minor contribution but I'm working on a couple theorems that have an existential form and I found this one to be useful.

jessealama committed:
refactor: simplify exists_not_mem_of_ne_top and use existing variables

- Use existing variables A, B from outer scope instead of introducing new ones
- Add autoparam for h_top to make the lemma easier to use
- Simplify proof using simpa as suggested by j-loreaux

j-loreaux reviewed (commented):
bors merge
